jetty-core/jetty-websocket/jetty-websocket-jetty-server/src/main/java/org/eclipse/jetty/websocket/server/ServerWebSocketContainer.java

```diff
@@ -45,11 +52,54 @@
  * <p>A server-side WebSocket container that allows to {@link #addMapping(String, WebSocketCreator) map}
  * URI paths to WebSocket endpoints and configure WebSocket parameters such as idle timeouts,
  * max WebSocket message sizes, etc.</p>
+ * <p>Direct WebSocket upgrades not mapped to URI paths are possible via
+ * {@link #upgrade(WebSocketCreator, Request, Response, Callback)}.</p>
  */
-public class ServerWebSocketContainer extends ContainerLifeCycle implements WebSocketContainer, Configurable, Invocable
+public class ServerWebSocketContainer extends ContainerLifeCycle implements WebSocketContainer, Configurable, Invocable, Request.Handler
 {
     private static final Logger LOG = LoggerFactory.getLogger(ServerWebSocketContainer.class);
 
+    /**
+     * <p>Returns the {@link ServerWebSocketContainer}, ensuring that
+     * it is available via {@link #from(Context)}.</p>
+     * <p>If the {@link ServerWebSocketContainer} is not already available,
+     * an instance is created, stored to be available via {@link #from(Context)}
+     * and returned.</p>
+     * <p>This method should be invoked during the setup of the
+     * {@link Handler} hierarchy.</p>
+     *
+     * @param server the {@link Server} object used to lookup common WebSocket components
+     * @param contextHandler the {@link ContextHandler} used to store the {@link ServerWebSocketContainer}
+     * @return a non-{@code null} {@link ServerWebSocketContainer}
+     */
+    public static ServerWebSocketContainer ensure(Server server, ContextHandler contextHandler)
+    {
+        Context context = contextHandler.getContext();
+        ServerWebSocketContainer container = from(context);
+        if (container == null)
+        {
+            WebSocketComponents components = WebSocketServerComponents.ensureWebSocketComponents(server, contextHandler);
+            WebSocketMappings mappings = new WebSocketMappings(components);
+            container = new ServerWebSocketContainer(mappings);
+            context.setAttribute(WebSocketContainer.class.getName(), container);
+        }
+        return container;
+    }
+
+    /**
+     * <p>Returns the {@link ServerWebSocketContainer} present as the context attribute
+     * under the name corresponding to the full qualified name of class
+     * {@link WebSocketContainer}.</p>
+     *
+     * @param context the {@link Context} to look for the attribute
+     * @return the {@link ServerWebSocketContainer} stored as an attribute,
+     * or {@code null} if no such attribute is present
+     */
+    public static ServerWebSocketContainer from(Context context)
```

Comment:
Maybe `get()` would be better than `from()`?
Seems like `from` implies it is creating something.